pkg/executor/bacalhau/bacalhau.go

```diff
@@ -126,121 +96,169 @@ func (executor *BacalhauExecutor) GetMachineSpecs() ([]data.MachineSpec, error)
 		}
 		specs = append(specs, spec)
 	}
+	if err != nil {
+		return nil, err
+	}
 	return specs, nil
 }
 
 func (executor *BacalhauExecutor) RunJob(
 	deal data.DealContainer,
 	module data.Module,
 ) (*executorlib.ExecutorResults, error) {
-	id, err := executor.getJobID(deal, module)
+	jobId, err := executor.getJobID(deal, module)
 	if err != nil {
 		return nil, err
 	}
 
-	jobState, err := executor.getJobState(deal.ID, id)
-	if err != nil {
-		return nil, err
-	}
+	var jobExecutions []*models.Execution
+	for {
+		jobInfo, err := executor.bacalhauClient.getJob(jobId)
+		if err != nil {
+			return nil, fmt.Errorf("error getting job %s: %s", jobId, err.Error())
+		}
+
+		if jobInfo.Executions == nil {
+			return nil, fmt.Errorf("no executions retrieved for job %s", jobId)
+		}
 
-	if len(jobState.State.Executions) <= 0 {
-		return nil, fmt.Errorf("no executions found for job %s", id)
+		jobExecutions = jobInfo.Executions.Items
+
+		if len(jobExecutions) > 0 {
+			if jobInfo.Job.State.StateType == models.JobStateTypeCompleted {
+				break
+			}
+		}
+
+		time.Sleep(time.Duration(executor.Options.JobStatusPollInterval) * time.Second)
 	}
 
-	// Check that the job completed successfully
-	if jobState.State.State != bacalhau.JobStateCompleted {
-		return nil, fmt.Errorf("job %s did not complete successfully: %s", id, jobState.State.State.String())
+	resultsDir, err := system.EnsureDataDir(filepath.Join(RESULTS_DIR, deal.ID))
+	if err != nil {
+		return nil, fmt.Errorf("error creating results directory: %s", err.Error())
+	}
+	outputDir, err := executor.fetchResults(jobId, resultsDir)
+	if err != nil {
+		return nil, fmt.Errorf("error fetching results: %s", err.Error())
 	}
 
-	system.EnsureDataDir(RESULTS_DIR)
-	resultsDir := system.GetDataDir(filepath.Join(RESULTS_DIR, deal.ID))
-	cidString := jobState.State.Executions[0].PublishedResult.CID
-	err = executor.ipfsClient.Get(context.Background(), cidString, resultsDir)
+	cid, err := executor.prepareResults(outputDir)
 	if err != nil {
-		return nil, fmt.Errorf("error getting results from IPFS %s -> %s", deal.ID, err)
+		return nil, fmt.Errorf("error preparing results: %s", err.Error())
 	}
 
-	// TODO: we should think about WASM and instruction count here
 	results := &executorlib.ExecutorResults{
-		ResultsDir:       resultsDir,
-		ResultsCID:       jobState.State.Executions[0].PublishedResult.CID,
+		ResultsDir:       outputDir,
+		ResultsCID:       cid,
 		InstructionCount: 1,
 	}
 
 	return results, nil
 }
 
-// run the bacalhau job and return the job ID
-func (executor *BacalhauExecutor) getJobID(
-	deal data.DealContainer,
-	module data.Module,
-) (string, error) {
-	// get a JSON string of the job
-	jsonBytes, err := json.Marshal(module.Job)
+func (executor *BacalhauExecutor) fetchResults(jobId string, resultsDir string) (string, error) {
+	resultsUrl, err := executor.bacalhauClient.getJobResult(jobId)
```

Comment:
nit: The Go convention is to capitalize all letters in initialisms like URL: https://go.dev/wiki/CodeReviewComments#initialisms